src/server/lib/okta/idx/interact.ts

```diff
@@ -0,0 +1,168 @@
+			if (oktaIdentityEngineSessionCookieId) {
+				await closeCurrentSession({
+					idx: oktaIdentityEngineSessionCookieId,
+				});
+			}
+		}
+
+		// generate the code verifier and code challenge for PKCE
+		// see https://www.oauth.com/oauth2-servers/pkce/authorization-request/
+		// the `code_verifier` is a cryptographically random string
+		// the `code_challenge` is the `code_verifier` hashed with SHA-256 and then base64url encoded
+		const codeVerifier = generators.codeVerifier(64);
+		const codeChallenge = generators.codeChallenge(codeVerifier);
+
+		// generate and store a "state"
+		// as a http only, secure, signed session cookie
+		// which is a json object that contains a stateParam and the query params
+		// the stateParam is used to protect against csrf
+		// we also set additional data in the state cookie that we need to persist
+		// between the authorization request and the callback
+		const authState = generateAuthorizationState(
+			getPersistableQueryParams(res.locals.queryParams),
+			confirmationPagePath,
+			doNotSetLastAccessCookie,
+			{
+				appPrefix: extraData?.appPrefix,
+				encryptedRegistrationConsents: extraData?.encryptedRegistrationConsents,
+				codeVerifier,
+			},
+		);
+		setAuthorizationStateCookie(authState, res);
+
+		// set up the authorization parameters to send to the /interact endpoint
+		// in order to start the interaction code flow
+		// see https://developer.okta.com/docs/concepts/interaction-code/
+		const authorizationParams: AuthorizationParameters = {
+			// the client_id is the id of the client application
+			client_id: okta.clientId,
+			// only use the interaction code flow callback uri, everything else is handled by standard authorization code flow
+			redirect_uri: ProfileOpenIdClientRedirectUris.INTERACTION_CODE,
+			// only authentication scopes for interaction code flow, everything else is handles by standard authorization code flow
+			scope: scopesForAuthentication.join(' '),
+			// we send the generated code challenge
+			code_challenge: codeChallenge,
+			// we use the S256 code challenge method, which is the only method supported
+			code_challenge_method: 'S256',
+			// we send the generated stateParam as the state parameter
+			state: authState.stateParam,
+		};
+
+		const authorizationSearchParams = new URLSearchParams(
+			authorizationParams as Record<string, string>,
+		);
+
+		const response = await fetch(
+			joinUrl(okta.orgUrl, '/oauth2/', okta.authServerId, '/v1/interact'),
+			{
+				method: 'POST',
+				headers: {
+					'Content-Type': 'application/x-www-form-urlencoded',
+				},
+				body: authorizationSearchParams.toString(),
+			},
+		);
+
+		if (!response.ok) {
+			// Check if the body is likely json using the content-type header
+			const contentType = response.headers.get('content-type');
+			if (contentType && contentType.includes('application/json')) {
+				// if so, parse the body as json, and handle the error
+				const error = await response.json();
```

Comment:
This _will_ throw in the unlikely scenario where `Content-Type` is JSON, but the body isn't (for instance, if it's an empty string). I would do something like:

```suggestion
				const error = response.json().then(j).catch((e) => new OAuthError(
					{
						error: 'unknown_error',
						error_description: await response.text(),
					},
					response.status,
				));
```

(This may not compile).